File: Leetcode/101.Symmetric_Tree/101.py
from typing import Optional
import queue
class TreeNode:
    def __init__(self,val=0,left=None,right=None):
        self.val=val
        self.left=left
        self.right=right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if root.left==None and root.right==None:
            return True
        que1=queue.Queue()
        que2=queue.Queue()
        que1.put(root.left)
        que2.put(root.right)
        while not que1.empty() or not que2.empty():
            size1=que1.qsize()
            size2=que2.qsize()
            if size1 != size2:
                return False
            for _ in range(size1):
                node1=que1.get()
                node2=que2.get()
                if node1.left:
                    que1.put(node1.left)
                if node1.right:
                    que1.put(node1.right)
                if node2.right and node1.left and (node2.right.val==node1.left.val):
                    que2.put(node2.right)
                elif node2.right and node1.left==None:
                    return False
                if node2.left and node1.right and (node2.left.val==node1.right.val):
                    que2.put(node2.left)
                elif node2.left and node1.right==None:
                    return False
        return True

sol=Solution()
root=TreeNode(1)
ptr=root
ptr.left=TreeNode(2)
ptr.right=TreeNode(2)
#ptr.left.left=TreeNode(3)
ptr.left.right=TreeNode(3)
#ptr.right.left=TreeNode(4)
ptr.right.right=TreeNode(5)
print("True") if sol.isSymmetric(root)==True else print("False")

File: Leetcode/101. Symmetric Tree/Java Script/101.js
function TreeNode(val = 0, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
}

var isSymmetric = function(root) {
    // 將左右子樹拆成兩個陣列分別記錄各個子結點
    let leftTree = [];
    let rightTree = [];
    if (root.left) {
        leftTree.push(root.left);
    }

    if (root.right) {
        rightTree.push(root.right);
    }
    
    if (leftTree.length != rightTree.length) {
        return false;
    }
    // 當左右子樹的長度相同時才繼續迴圈
    while (leftTree.length > 0 && rightTree.length > 0) {
        let leftSize = leftTree.length;
        let rightSize = rightTree.length;
        // 左右子樹長度不同表示非完全對稱
        if (leftSize != rightSize) {
            return false;
        }
        // 左右子樹值不同表示非完全對稱
        for (let i = 0 ; i < leftSize ; i++) {
            if (leftTree[i].val != rightTree[i].val) {
                return false;
            }
        }
        
        for (let i = 0 ; i < leftSize ; i++) {
            let leftNode = leftTree.shift();
            let rightNode = rightTree.shift();
            // 左右子樹分別遍歷各節點，這邊注意因為要取頭節點，所以左右子樹要從相反方向遍歷
            if (leftNode.left && rightNode.right) {
                leftTree.push(leftNode.left);
                rightTree.push(rightNode.right);
            } else if ((!leftNode.left && rightNode.right) || (leftNode.left && !rightNode.right)) { // 判斷左右子樹是否對稱
                return false;
            }
             // 左右子樹分別遍歷各節點，這邊注意因為要取頭節點，所以左右子樹要從相反方向遍歷
            if (leftNode.right && rightNode.left) {
                leftTree.push(leftNode.right);
                rightTree.push(rightNode.left);
            } else if ((!leftNode.right && rightNode.left) || (leftNode.right && !rightNode.left)) { // 判斷左右子樹是否對稱
                return false;
            }
        }
    }

    return true;
};


// let root = new TreeNode(1, new TreeNode(2, new TreeNode(3), new TreeNode(4)), new TreeNode(2, new TreeNode(4), new TreeNode(3)));
let root = new TreeNode(1, new TreeNode(2, null, new TreeNode(3)), new TreeNode(2, null, new TreeNode(3)));

console.log(isSymmetric(root));
